### PR TITLE
cuda-12.8 and FlashAttention 2.x compatibility 

### PR DESCRIPTION
**Pull request by lidanwu**
This PR is aimed to support modern GPU environment. 

- Includes compatibility updates for both FlashAttention 1.x and 2.x. 
  - In particular, it supports CUDA 12.8 and flash-attn 2.8.3 for modern NVIDIA Blackwell GPUs while retaining backward compatibility with FA1-style checkpoints through parameter-name mapping during checkpoint loading.
  - This is done by rewriting a high-level wrapper for FlashHMA using FA2-backend but in FA1-style & logic (see `scgpt/model/flash_attn_compat.py`)
  - `scgpt.utils.load_pretrained` is also upated to support loading of FA1/FA2 trained checkpoint into new environment using any backend (FA1/FA2/pyTorch).
  - For quick backend and memory-path checks in your environment, use the diagnostic tool at `scripts/diagnose_flash_attn.py`

- Adds pure-python `Vocab` since the development of `torchtext` has stopped with its latest version at 0.18.0 and `torchtext >=0.14` ships `Vocab` as a C++ extension type that cannot be subclassed like before.

- Adds more tests and correct legacy issues.

- bump to `0.2.5` and update `pyproject.toml` to standard PEP 621 + setuptools so installs/builds no longer depend on Poetry tooling.

The operation and model inference outcomes has been verified in two different linux environments. 

1. `cuda-11.7` environment that uses `torch 2.0.1+cu117`, `flash-attn 1.0.4`, `torchtext 0.15.2+cpu`
image

2. `cuda-12.8` environment that uses `torch 2.11.0+cu128`, `flash-attn 2.8.3` but no `torchtext`
image

Confirmed that updated package in this PR gave same model inference results (confusion matrix of annotation outcomes) as those using the offical `scgpt 0.2.4` when running in `cuda-11.7` environment. The `cuda-11.7` results are also consistent with the ones using updated package in `cuda-12.8` environment.

### CODE REVIEW

**Comment by lidanwu** [comment]
@subercui First of all, thanks for developing the scGPT algorithm and making this repository public. 
I made this PR to enable training and inference using `flash-attn` in `cuda-12.8` environment such that I could take advantage of the NVIDIA Blackwell GPU. I did all the tests I could think of and confirmed it's working as expected on my ends. I thought this might be useful for other developers too and would appreciate your team's review on it.

**Comment by subercui** [comment]
> @subercui First of all, thanks for developing the scGPT algorithm and making this repository public. I made this PR to enable training and inference using `flash-attn` in `cuda-12.8` environment such that I could take advantage of the NVIDIA Blackwell GPU. I did all the tests I could think of and confirmed it's working as expected on my ends. I thought this might be useful for other developers too and would appreciate your team's review on it.

Thank you! Reviewing this right now

**Comment by subercui** [comment]
@codex review

**Review by subercui — APPROVED** [review]
